File: frontend/src/lib/buildplan.js
// A whole build planned at once on the Build planner tab: one row per slot
// (or, without the slot table, per engineerable module), and the pure
// shaping between those rows, what is saved per ship, and what the backend
// is asked for. No Svelte here so it can be tested.

const TOP_GRADE = 5;
/** The swap that clears a slot (the backend's `ed_ships::EMPTY`). */
export const EMPTY = "empty";

/**
 * One plan row per slot when the slot table is given (every slot of the
 * hull, fitted or empty, so a module can be swapped in), else one per
 * engineerable module. A saved plan (by slot) wins over the defaults; the
 * defaults are the fitted blueprint continued to the top grade, included
 * when there is something left to do, and an unengineered module left out
 * until a blueprint is picked. A saved swap puts the new module in the row:
 * it starts at grade 0 with the new module's type.
 * @param {Array<{slot:string, slot_name:string, item?:string, item_name:string, module_type?:string|null, blueprint?:string|null, grade?:number|null}>} modules
 * @param {Record<string, {blueprint?:string, target_grade?:number, experimental?:string, include?:boolean, swap?:string|null}>} saved
 * @param {Array<{slot:string, slot_name:string, group:string, size:number, fitted?:string|null, fitted_name?:string|null, candidates:Array<{item:string, item_name:string, module_type?:string|null}>}>|null} slots
 */
export function planRows(modules, saved = {}, slots = null) {
  const bySlot = new Map((modules ?? []).map((m) => [m.slot.toLowerCase(), m]));
  const shape = (m, slotInfo) => {
    const slot = slotInfo?.slot ?? m.slot;
    const s = saved?.[slot] ?? {};
    const fittedGrade = m?.grade ?? 0;
    const fittedBlueprint = m?.blueprint ?? "";
    const emptied = s.swap === EMPTY && (slotInfo?.can_empty ?? true);
    const candidate = s.swap && !emptied ? findCandidate(slotInfo?.candidates ?? [], s.swap, s.preset ?? null) : null;
    const swap = emptied ? EMPTY : candidate ? candidate.item : null;
    const preset = candidate?.preset ?? null;
    // A pre-engineered module comes with its grade; a plain swap starts at 0.
    const from = swap ? (candidate?.preset_grade ?? 0) : fittedGrade;
    const blueprint = swap ? (candidate?.preset_blueprint ?? s.blueprint ?? "") : s.blueprint ?? fittedBlueprint;
    const module_type = emptied ? null : swap ? candidate.module_type ?? null : m?.module_type ?? null;
    return {
      slot,
      slot_name: slotInfo?.slot_name ?? m.slot_name,
      group: slotInfo?.group ?? null,
      size: slotInfo?.size ?? null,
      fitted_item: m?.item ?? slotInfo?.fitted ?? null,
      fitted_name: m?.item_name ?? slotInfo?.fitted_name ?? null,
      fitted_grade: fittedGrade,
      fitted_blueprint: fittedBlueprint,
      fitted_module_type: m?.module_type ?? null,
      swap,
      preset,
      item: emptied ? null : swap ?? m?.item ?? slotInfo?.fitted ?? null,
      item_name: emptied ? null : swap ? candidate.item_name : m?.item_name ?? slotInfo?.fitted_name ?? null,
      module_type,
      blueprint: module_type ? blueprint : "",
      from_grade: from,
      target_grade: Number(s.target_grade ?? TOP_GRADE),
      experimental: module_type ? s.experimental ?? "" : "",
      include: module_type ? s.include ?? (Boolean(blueprint) && from < TOP_GRADE) : false,
    };
  };
  if (slots) return slots.map((si) => shape(bySlot.get(si.slot.toLowerCase()) ?? null, si));
  return (modules ?? []).filter((m) => m.module_type).map((m) => shape(m, null));
}

/** The candidate with this item and (when given) this pre-engineered variant. */
export function findCandidate(candidates, item, preset = null) {
  const want = String(item).toLowerCase();
  return candidates.find((c) => c.item.toLowerCase() === want && (c.preset ?? null) === (preset ?? null)) ?? null;
}

/** The key a swap dropdown option carries: item, and the variant when there is one. */
export const swapKey = (item, preset = null) => (item ? `${item}|${preset ?? ""}` : "");

/**
 * A row with another module swapped into its slot (a candidate from the
 * slot table), emptied (`candidate.item === EMPTY`), or, with `candidate`
 * null or the fitted module itself, back to what is fitted. A swap starts
 * over: grade 0, no blueprint, not included until one is picked — except
 * a pre-engineered module, which comes at its grade with its blueprint,
 * so only an experimental is left to plan.
 */
export function withSwap(row, candidate) {
  if (!candidate || (row.fitted_item && !candidate.preset && candidate.item.toLowerCase() === row.fitted_item.toLowerCase())) {
    const blueprint = row.fitted_blueprint ?? "";
    const type = row.fitted_module_type ?? null;
    return { ...row, swap: null, preset: null, item: row.fitted_item, item_name: row.fitted_name, module_type: type, blueprint: type ? blueprint : "", from_grade: row.fitted_grade ?? 0, target_grade: TOP_GRADE, experimental: "", include: Boolean(type) && Boolean(blueprint) && (row.fitted_grade ?? 0) < TOP_GRADE };
  }
  if (candidate.item === EMPTY) {
    return { ...row, swap: EMPTY, preset: null, item: null, item_name: null, module_type: null, blueprint: "", from_grade: 0, target_grade: TOP_GRADE, experimental: "", include: false };
  }
  const from = candidate.preset_grade ?? 0;
  return { ...row, swap: candidate.item, preset: candidate.preset ?? null, item: candidate.item, item_name: candidate.item_name, module_type: candidate.module_type ?? null, blueprint: candidate.preset_blueprint ?? "", from_grade: from, target_grade: TOP_GRADE, experimental: "", include: false };
}

/** The swaps the rows carry, shaped for the backend. */
export function swapsFrom(rows) {
  return rows.filter((r) => r.swap).map((r) => (r.preset ? { slot: r.slot, item: r.swap, preset: r.preset } : { slot: r.slot, item: r.swap }));
}

/** Copy one row's choices onto every row of the same module type ("all 9 pulse lasers"). */
export function sameForAll(rows, row) {
  return rows.map((r) =>
    r.module_type && r.module_type === row.module_type
      ? { ...r, blueprint: row.blueprint, target_grade: row.target_grade, experimental: row.experimental, include: row.include && hasWork({ ...r, blueprint: row.blueprint, target_grade: row.target_grade, experimental: row.experimental }) }
      : r,
  );
}

/** How many rows share each module type, for the "same for all N" button. */
export function groupCounts(rows) {
  const counts = new Map();
  for (const r of rows) if (r.module_type) counts.set(r.module_type, (counts.get(r.module_type) ?? 0) + 1);
  return counts;
}

/** The blueprint still has grades to roll (a module at the top grade has none). */
export function rollsLeft(row) {
  return Boolean(row.blueprint) && Number(row.target_grade) > row.from_grade;
}

/** Something to do: grades to roll, or an experimental to apply (a top-grade module can still take one). */
export function hasWork(row) {
  return rollsLeft(row) || Boolean(row.experimental);
}

/** A row is planned when it is included and has work. */
export function isPlanned(row) {
  return Boolean(row.include) && hasWork(row);
}

/** What the backend is asked for: the planned rows, shaped as PlanItem. */
export function planRequest(rows) {
  return rows.filter((r) => r.module_type && isPlanned(r)).map((r) => ({
    slot: r.slot,
    module_type: r.module_type,
    // A top-grade module with only an experimental planned sends no blueprint: nothing to roll.
    blueprint: rollsLeft(r) ? r.blueprint : null,
    from_grade: r.from_grade,
    target_grade: Number(r.target_grade),
    experimental: r.experimental || null,
  }));
}

/** The planned blueprints at their target grade, for the SLEF export. */
export function proposedFor(rows) {
  return rows
    .filter((r) => r.module_type && isPlanned(r) && rollsLeft(r))
    .map((r) => ({ slot: r.slot, module_type: r.module_type, blueprint: r.blueprint, grade: Number(r.target_grade) }));
}

/** What is saved per ship: the choices, by slot (a swap only when there is one). */
export function savedFrom(rows) {
  const out = {};
  for (const r of rows) {
    const s = { blueprint: r.blueprint, target_grade: Number(r.target_grade), experimental: r.experimental, include: r.include };
    if (r.swap) s.swap = r.swap;
    if (r.preset) s.preset = r.preset;
    out[r.slot] = s;
  }
  return out;
}

/**
 * Slot names folded into ranges: ["Large hardpoint 1", "Large hardpoint 2",
 * "Large hardpoint 4", "Utility 1"] → "Large hardpoints 1–2, 4; Utility 1".
 * One line for nine lasers instead of nine lines (maintainer, 2026-09-19:
 * "seems a bit spammy and repeated").
 */
export function compactSlots(names) {
  const groups = new Map();
  for (const n of names) {
    const m = /^(.*?)\s*(\d+)$/.exec(n);
    if (!m) { groups.set(n, null); continue; }
    if (!groups.has(m[1])) groups.set(m[1], []);
    groups.get(m[1]).push(Number(m[2]));
  }
  const parts = [];
  for (const [prefix, nums] of groups) {
    if (!nums) { parts.push(prefix); continue; }
    nums.sort((a, b) => a - b);
    const ranges = [];
    let start = nums[0], prev = nums[0];
    for (const n of nums.slice(1)) {
      if (n === prev + 1) { prev = n; continue; }
      ranges.push(start === prev ? `${start}` : `${start}–${prev}`);
      start = prev = n;
    }
    ranges.push(start === prev ? `${start}` : `${start}–${prev}`);
    // "Large hardpoints 1–4" reads right; "Utilitys" does not: only the word that takes an s gets one.
    const plural = nums.length > 1 && /hardpoint$/i.test(prefix) ? `${prefix}s` : prefix;
    parts.push(`${plural} ${ranges.join(", ")}`);
  }
  return parts.join("; ");
}

const jobKey = (it) => `${it.blueprint} G${it.target_grade}`;

/** The itinerary, one line per engineer, each job grouped by blueprint and grade. */
export function itinerary(report) {
  return (report?.engineers ?? []).map((stop) => {
    const mine = report.items.filter((it) => it.assigned_to === stop.engineer);
    const groups = new Map();
    for (const it of mine) {
      const k = jobKey(it);
      if (!groups.has(k)) groups.set(k, { what: k, module_type: it.module_type, slots: [] });
      groups.get(k).slots.push(it.slot_name);
    }
    const jobs = [...groups.values()]
      .map((g) => ({ what: g.what, module_type: g.module_type, count: g.slots.length, slots: compactSlots(g.slots) }))
      .sort((a, b) => b.count - a.count || a.what.localeCompare(b.what));
    return { engineer: stop.engineer, rank: stop.rank, jobs };
  });
}

/**
 * What no unlocked engineer can do at the asked grade, grouped, with the
 * way out: who takes it part-way today, and who to unlock for the rest.
 */
export function blocked(report) {
  const groups = new Map();
  for (const it of report?.items ?? []) {
    if (!it.blueprint || it.reachable) continue;
    const k = `${it.module_type}|${jobKey(it)}`;
    if (!groups.has(k)) {
      groups.set(k, {
        what: jobKey(it),
        module_type: it.module_type,
        target_grade: it.target_grade,
        max_reachable_grade: it.max_reachable_grade ?? null,
        today: it.engineers.filter((e) => e.unlocked && it.max_reachable_grade && e.max_grade >= it.max_reachable_grade).map((e) => e.engineer),
        unlock: it.engineers.filter((e) => !e.unlocked && e.max_grade >= it.target_grade).map((e) => ({ engineer: e.engineer, status: e.status })),
        slots: [],
      });
    }
    groups.get(k).slots.push(it.slot_name);
  }
  return [...groups.values()].map((g) => ({ ...g, count: g.slots.length, slots: compactSlots(g.slots) }));
}

const emptyRow = (slot, slot_name) => ({ slot, slot_name, group: null, size: null, fitted_item: null, fitted_name: null, fitted_grade: 0, fitted_blueprint: "", fitted_module_type: null, swap: null, preset: null, item: null, item_name: null, module_type: null, blueprint: "", from_grade: 0, target_grade: TOP_GRADE, experimental: "", include: false });

/**
 * An imported build laid over the rows: every module the build has where
 * the ship has something else becomes that row's swap; every engineered
 * module of the build becomes that slot's plan (a swapped module starts
 * from grade 0, the same blueprint continues from the fitted grade, a
 * module the ship already has is left unticked); a slot the rows do not
 * have gets one; and everything the build does not engineer is unticked.
 */
export function applyImport(rows, imported) {
  const bySlot = new Map((imported?.items ?? []).map((it) => [it.slot, it]));
  const swapBySlot = new Map((imported?.swaps ?? []).map((sw) => [sw.slot, sw]));
  const lay = (r, it, sw) => {
    const emptied = sw?.want_item === EMPTY;
    const swapped = emptied
      ? { swap: EMPTY, preset: null, item: null, item_name: null, module_type: null, from_grade: 0 }
      : sw ? { swap: sw.want_item, preset: sw.preset ?? null, item: sw.want_item, item_name: sw.want, module_type: it?.module_type ?? null, from_grade: 0 } : {};
    if (!it) return { ...r, ...swapped, blueprint: "", experimental: "", include: false };
    return {
      ...r,
      ...swapped,
      item_name: it.item_name,
      module_type: it.module_type,
      blueprint: it.blueprint ?? "",
      from_grade: it.from_grade,
      target_grade: it.target_grade,
      experimental: it.experimental ?? "",
      include: !it.done,
    };
  };
  const out = rows.map((r) => lay(r, bySlot.get(r.slot), swapBySlot.get(r.slot)));
  const have = new Set(rows.map((r) => r.slot));
  for (const slot of new Set([...bySlot.keys(), ...swapBySlot.keys()])) {
    if (have.has(slot)) continue;
    const it = bySlot.get(slot);
    const sw = swapBySlot.get(slot);
    out.push(lay(emptyRow(slot, it?.slot_name ?? sw?.slot_name ?? slot), it, sw));
  }
  return out;
}

File: src-tauri/src/build_plan.rs
//! A whole build planned at once: every engineerable slot with a chosen
//! blueprint, grade and experimental, reported as ONE material list, one
//! shopping list and one engineer itinerary (maintainer, 2026-09-19: "my
//! type 10 has 9 weapon hardpoints — I'd like to be able to plan out all 9
//! at once and get the list").
//!
//! The per-module machinery already existed (gap reports, engineer access,
//! the shopping list from what the commander carries); this module pools
//! it across slots. Materials pool by name, so nine Focused pulse lasers
//! need nine times the iron — the number a commander actually has to
//! collect, which no per-module view can show.

use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, HashMap};

use crate::commands::{self, EngineerAccess, ShoppingReport};
use crate::state::AppState;
use ed_engineering::RequirementLine;

fn five() -> i64 {
    5
}

/// One slot's plan, as the Ships tab sends it.
#[derive(Debug, Clone, Deserialize)]
pub struct PlanItem {
    /// The journal's `Slot` value ("LargeHardpoint1").
    pub slot: String,
    /// Blueprint-data module type ("Pulse Laser").
    pub module_type: String,
    /// Blueprint name ("Focused Weapon"); None when only an experimental is planned.
    #[serde(default)]
    pub blueprint: Option<String>,
    /// The grade already on the module (0 when unengineered).
    #[serde(default)]
    pub from_grade: i64,
    #[serde(default = "five")]
    pub target_grade: i64,
    #[serde(default)]
    pub experimental: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct PlanItemReport {
    pub slot: String,
    pub slot_name: String,
    pub item_name: String,
    pub module_type: String,
    pub blueprint: Option<String>,
    pub from_grade: i64,
    pub target_grade: i64,
    pub experimental: Option<String>,
    /// This slot's own materials (blueprint rolls plus the experimental).
    pub lines: Vec<RequirementLine>,
    /// Every engineer who works the blueprint, with their cap and status.
    pub engineers: Vec<EngineerAccess>,
    /// Someone unlocked offers the target grade.
    pub reachable: bool,
    pub max_reachable_grade: Option<i64>,
    /// The engineer this slot is routed to in the itinerary, if any.
    pub assigned_to: Option<String>,
}

/// One stop on the itinerary: an unlocked engineer and the slots to bring.
#[derive(Debug, Serialize, PartialEq, Eq)]
pub struct PlanEngineer {
    pub engineer: String,
    pub rank: Option<i64>,
    pub jobs: Vec<String>,
}

/// A swapped-in module a technology broker sells: its unlock, broken
/// down (maintainer, 2026-09-20: "refer to the recipe for the component
/// and break it down that way"). Materials pool into the plan's list;
/// commodities are bought at a market and listed apart.
#[derive(Debug, Serialize)]
pub struct Unlock {
    pub slot_name: String,
    pub item_name: String,
    /// "Guardian" or "Human": which technology broker.
    pub broker: String,
    pub materials: Vec<RequirementLine>,
    pub commodities: Vec<(String, i64)>,
}

#[derive(Debug, Serialize)]
pub struct BuildPlanReport {
    pub ship: String,
    pub ship_name: Option<String>,
    pub items: Vec<PlanItemReport>,
    /// Every material the whole plan needs, pooled, against the inventory.
    pub materials: Vec<RequirementLine>,
    pub fully_met: bool,
    /// Fewest engineers that cover every reachable slot, each with its slots.
    pub engineers: Vec<PlanEngineer>,
    /// Slots no unlocked engineer can take to the asked grade.
    pub unassigned: Vec<String>,
    /// Trades and farming for the pooled shortfall; None when nothing is short.
    pub shopping: Option<ShoppingReport>,
    /// Technology-broker unlocks the swaps need, materials already pooled above.
    pub unlocks: Vec<Unlock>,
}

/// Materials pooled by name across every slot. `have` is the same for all
/// lines of one material, so the pooled line keeps it.
pub fn pool(per_item: &[Vec<RequirementLine>]) -> Vec<RequirementLine> {
    let mut by_name: BTreeMap<String, (i64, i64)> = BTreeMap::new();
    for lines in per_item {
        for l in lines {
            let e = by_name.entry(l.material.clone()).or_insert((0, l.have));
            e.0 += l.need;
            e.1 = l.have;
        }
    }
    by_name
        .into_iter()
        .map(|(material, (need, have))| RequirementLine { material, need, have })
        .collect()
}

/// A slot that can be done by these engineers (unlocked ones offering the
/// asked grade), labelled for the itinerary.
pub struct Job {
    pub label: String,
    pub candidates: Vec<(String, Option<i64>)>,
}

/// Fewest stops: greedy set cover. Each round takes the engineer who can
/// still do the most remaining slots (ties to the name, so two reads give
/// the same itinerary); slots nobody can do are reported, never dropped.
pub fn assign(jobs: &[Job]) -> (Vec<PlanEngineer>, Vec<String>) {
    let mut remaining: Vec<usize> = (0..jobs.len()).collect();
    let mut out: Vec<PlanEngineer> = Vec::new();
    loop {
        let mut counts: BTreeMap<&str, (usize, Option<i64>)> = BTreeMap::new();
        for &i in &remaining {
            for (name, rank) in &jobs[i].candidates {
                let e = counts.entry(name.as_str()).or_insert((0, *rank));
                e.0 += 1;
            }
        }
        // Most slots first; BTreeMap order breaks ties alphabetically.
        let Some((best, (n, rank))) = counts.into_iter().max_by(|a, b| a.1 .0.cmp(&b.1 .0).then(b.0.cmp(a.0))) else {
            break;
        };
        if n == 0 {
            break;
        }
        let (mine, rest): (Vec<usize>, Vec<usize>) =
            remaining.iter().partition(|&&i| jobs[i].candidates.iter().any(|(c, _)| c == best));
        out.push(PlanEngineer { engineer: best.to_string(), rank, jobs: mine.iter().map(|&i| jobs[i].label.clone()).collect() });
        remaining = rest;
        if remaining.is_empty() {
            break;
        }
    }
    let unassigned = remaining.iter().map(|&i| jobs[i].label.clone()).collect();
    (out, unassigned)
}

/// The report, everything but the trader lookup (which is async and done
/// by the command).
pub fn report(state: &AppState, ship_id: Option<i64>, items: &[PlanItem], swaps: &[(String, String)], minimum: bool, complete: bool) -> Result<BuildPlanReport, String> {
    if items.is_empty() && swaps.is_empty() {
        return Err("nothing planned: pick a blueprint for at least one module".into());
    }
    let loadout = commands::ship_loadout(state, ship_id)?;
    let (have, engineers) = state.with_read(|s| {
        (
            commands::material_inventory(s.conn()),
            ed_store::query::engineers(s.conn()).unwrap_or_default(),
        )
    });
    let catalog = &state.engineering;

    let mut reports: Vec<PlanItemReport> = Vec::new();
    let mut per_item: Vec<Vec<RequirementLine>> = Vec::new();
    let mut jobs: Vec<Job> = Vec::new();
    for it in items {
        let module = loadout.modules.iter().find(|m| m.slot == it.slot);
        let (slot_name, item_name) = match module {
            Some(m) => (m.slot_name.clone(), m.item_name.clone()),
            None => (it.slot.clone(), String::new()),
        };
        let label = format!("{slot_name}: {}", it.blueprint.as_deref().unwrap_or("experimental only"));
        let mut lines: Vec<RequirementLine> = Vec::new();
        let mut access: Vec<EngineerAccess> = Vec::new();
        let (mut reachable, mut max_reachable) = (false, None);
        if let Some(bp) = &it.blueprint {
            if it.target_grade <= it.from_grade {
                return Err(format!("{slot_name}: target grade {} is not above the fitted grade {}", it.target_grade, it.from_grade));
            }
            let gap = if minimum {
                catalog.gap_report(&it.module_type, bp, it.from_grade, it.target_grade, &have)
            } else {
                catalog.gap_report_realistic(&it.module_type, bp, it.from_grade, it.target_grade, complete, &have)
            };
            if gap.lines.is_empty() && catalog.find(&it.module_type, bp, it.target_grade).is_none() {
                return Err(format!("{slot_name}: no blueprint {bp:?} at grade {} for {:?}", it.target_grade, it.module_type));
            }
            lines.extend(gap.lines);
            if let Some((a, r, m)) = commands::access_for(&engineers, catalog, &it.module_type, bp, it.target_grade) {
                access = a;
                reachable = r;
                max_reachable = m;
            }
            jobs.push(Job {
                label: format!("{label} G{}", it.target_grade),
                candidates: access
                    .iter()
                    .filter(|a| a.unlocked && a.max_grade >= it.target_grade)
                    .map(|a| (a.engineer.clone(), a.rank))
                    .collect(),
            });
        }
        if let Some(x) = &it.experimental {
            let gap = catalog
                .experimental_gap(&it.module_type, x, &have)
                .ok_or_else(|| format!("{slot_name}: no experimental effect {x:?} for {:?}", it.module_type))?;
            lines.extend(gap.lines);
        }
        per_item.push(lines.clone());
        reports.push(PlanItemReport {
            slot: it.slot.clone(),
            slot_name,
            item_name,
            module_type: it.module_type.clone(),
            blueprint: it.blueprint.clone(),
            from_grade: it.from_grade,
            target_grade: it.target_grade,
            experimental: it.experimental.clone(),
            lines,
            engineers: access,
            reachable,
            max_reachable_grade: max_reachable,
            assigned_to: None,
        });
    }

    // Swaps to technology-broker modules: the unlock recipe, materials
    // pooled with the rest, commodities listed to buy.
    let jc = ed_journal::Catalog::load();
    let mut unlocks = Vec::new();
    for (slot, item) in swaps {
        let item_name = ed_journal::modules::item_name(item);
        let Some(recipe) = catalog.unlock_recipe(&item_name) else { continue };
        let mut mats = Vec::new();
        let mut comms = Vec::new();
        for ing in &recipe.ingredients {
            match jc.by_name(&ing.name) {
                Some(i) if i.kind == ed_journal::Kind::Commodity => comms.push((ing.name.clone(), ing.quantity)),
                _ => mats.push(RequirementLine { material: ing.name.clone(), need: ing.quantity, have: *have.get(&ing.name).unwrap_or(&0) }),
            }
        }
        per_item.push(mats.clone());
        unlocks.push(Unlock {
            slot_name: ed_journal::modules::slot_name(slot),
            item_name,
            broker: recipe.module_type.clone(),
            materials: mats,
            commodities: comms,
        });
    }
    let materials = pool(&per_item);
    let fully_met = materials.iter().all(|l| l.have >= l.need);
    let (itinerary, unassigned) = assign(&jobs);
    for stop in &itinerary {
        for r in reports.iter_mut().filter(|r| r.blueprint.is_some()) {
            let label = format!("{}: {} G{}", r.slot_name, r.blueprint.as_deref().unwrap_or(""), r.target_grade);
            if stop.jobs.contains(&label) {
                r.assigned_to = Some(stop.engineer.clone());
            }
        }
    }

    let shopping = if fully_met {
        None
    } else {
        let need: HashMap<String, i64> = materials.iter().map(|l| (l.material.clone(), l.need)).collect();
        let galaxy = state.routing.galaxy(&state.data_dir);
        Some(state.with_read(|s| commands::shopping_from_need(s.conn(), galaxy.as_deref(), need, have.clone()))?)
    };

    Ok(BuildPlanReport {
        ship: loadout.ship,
        ship_name: loadout.ship_name,
        items: reports,
        materials,
        fully_met,
        engineers: itinerary,
        unassigned,
        shopping,
        unlocks,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn line(m: &str, need: i64, have: i64) -> RequirementLine {
        RequirementLine { material: m.into(), need, have }
    }

    /// Nine pulse lasers need nine times the iron: the pooled line is the
    /// number to collect, and "have" is the one inventory, not nine copies.
    #[test]
    fn materials_pool_by_name_across_slots() {
        let per_item = vec![
            vec![line("Iron", 3, 10), line("Nickel", 2, 0)],
            vec![line("Iron", 3, 10)],
            vec![line("Iron", 3, 10), line("Nickel", 1, 0)],
        ];
        let pooled = pool(&per_item);
        assert_eq!(pooled.len(), 2);
        assert_eq!((pooled[0].material.as_str(), pooled[0].need, pooled[0].have), ("Iron", 9, 10));
        assert_eq!((pooled[1].material.as_str(), pooled[1].need, pooled[1].have), ("Nickel", 3, 0));
    }

    fn job(label: &str, who: &[&str]) -> Job {
        Job { label: label.into(), candidates: who.iter().map(|w| (w.to_string(), Some(5))).collect() }
    }

    /// The itinerary is the fewest stops: one engineer who covers eight
    /// lasers and the distributor beats visiting two, and the slot only
    /// one engineer can do still gets its stop. A slot nobody unlocked
    /// can do is named, not dropped.
    #[test]
    fn the_itinerary_takes_the_fewest_engineers_and_names_what_nobody_can_do() {
        let mut jobs: Vec<Job> = (1..=8)
            .map(|i| job(&format!("Large Hardpoint {i}: Focused Weapon G4"), &["The Dweller", "Broo Tarquin"]))
            .collect();
        jobs.push(job("Power Distributor: Charge Enhanced G5", &["The Dweller"]));
        jobs.push(job("Frame Shift Drive: Increased FSD Range G5", &["Felicity Farseer"]));
        jobs.push(job("Thrusters: Dirty Drive Tuning G5", &[]));
        let (stops, unassigned) = assign(&jobs);
        assert_eq!(stops.iter().map(|s| (s.engineer.as_str(), s.jobs.len())).collect::<Vec<_>>(), [("The Dweller", 9), ("Felicity Farseer", 1)]);
        assert_eq!(unassigned, ["Thrusters: Dirty Drive Tuning G5"]);
    }

    /// Equal coverage falls to the name, so two reads give one itinerary.
    #[test]
    fn ties_break_by_name() {
        let jobs = vec![job("a", &["Zed", "Abe"]), job("b", &["Zed", "Abe"])];
        let (stops, _) = assign(&jobs);
        assert_eq!(stops.len(), 1);
        assert_eq!(stops[0].engineer, "Abe");
    }
}
